mlir/lib/IR/AffineExpr.cpp

```diff
@@ -1178,9 +1178,9 @@ static AffineExpr getSemiAffineExprFromFlatForm(ArrayRef<int64_t> flatExprs,
       continue;
     AffineExpr expr = it.value();
     auto binaryExpr = dyn_cast<AffineBinaryOpExpr>(expr);
-    if (!binaryExpr)
-      continue;
-
+    assert(binaryExpr &&
```

Comment:
Instead of dyn_cast + assert, use `cast` and add the assertion message as a comment on top.

Reply:
Yes, that's exactly what I meant. The assertion message (string) can be added as a comment so developers know better the reason for using `cast`.